libc/socket/unix-socket.c

```diff
@@ -904,6 +933,399 @@ TEST(test_unix_socket, poll)
+
+	fds[2].fd = accept(fds[0].fd, NULL, NULL);
+	fds[2].events = POLLIN;
+	TEST_ASSERT(fds[2].fd > 0);
+
+	TEST_ASSERT_EQUAL_INT(1, poll(fds, 2, 1000));
+	TEST_ASSERT_EQUAL_INT(0, fds[0].revents);
+	TEST_ASSERT_EQUAL_INT(POLLOUT, fds[1].revents);
+
+	send_msg(fds[1].fd, 1, 0);
+
+	TEST_ASSERT_EQUAL_INT(2, poll(fds, 3, 1000));
+	TEST_ASSERT_EQUAL_INT(0, fds[0].revents);
+	TEST_ASSERT_EQUAL_INT(POLLOUT, fds[1].revents);
+	TEST_ASSERT_EQUAL_INT(POLLIN, fds[2].revents);
+
+	read_msg(fds[2].fd, 1, 0);
+
+	close(fds[0].fd);
+	close(fds[1].fd);
+	close(fds[2].fd);
+}
+
+
+TEST(test_unix_socket, accept_connect_async)
+{
+	unsigned int i;
+
+	for (i = 0; i < CONNECTED_LOOP_CNT; ++i) {
+		unix_accept_connect_async(SOCK_STREAM);
+		unix_accept_connect_async(SOCK_SEQPACKET);
+	}
+}
+
+
+static void unix_accept_connect_liveness_helper(int type)
+{
+	pid_t pid;
+	int fd, named, rv, conn;
+	struct pollfd fds[2];
+	int status;
+
+	const char *socket_name = "/tmp/test_accept_connect";
+
+	/* blocking connect, blocking accept */
+
+	pid = checked_fork();
+
+	if (pid) {
+		if ((named = unix_named_socket(type, socket_name)) < 0)
+			FAIL("unix_named_socket");
+
+		if (listen(named, 0) < 0)
+			FAIL("listen");
+
+		if ((conn = accept(named, NULL, NULL)) < 0)
+			FAIL("accept");
+
+		read_msg(conn, pid, 0);
+
+		if (waitpid(pid, &status, 0) < 0)
+			FAIL("waitpid");
+
+		TEST_ASSERT(WIFEXITED(status));
+		TEST_ASSERT_EQUAL_INT(0, WEXITSTATUS(status));
+
+		close(conn);
+		close(named);
+	}
+	else {
+		if ((fd = socket(AF_UNIX, type, PF_UNSPEC)) < 0)
+			exit(1);
+
+		while (true) {
+			rv = connect_to_named(fd, socket_name);
+			if (rv == 0)
+				break;
+			else {
+				CHILD_ASSERT_ERRNO(rv, ECONNREFUSED);
+				usleep(150);
+			}
+		}
+
+		send_msg(fd, pid, 0);
+
+		close(fd);
+
+		exit(0);
+	}
+
+	/* blocking connect, nonblocking accept */
+
+	pid = checked_fork();
+
+	if (pid) {
+		if ((named = unix_named_socket(type, socket_name)) < 0)
+			FAIL("unix_named_socket");
+
+		if (set_nonblock(named, 1) < 0)
+			FAIL("set_nonblock");
+
+		if (listen(named, 0) < 0)
+			FAIL("listen");
+
+		fds[0].fd = named;
+		fds[0].events = POLLIN;
+
+		TEST_ASSERT_EQUAL_INT(1, poll(fds, 1, 500));
+		TEST_ASSERT_EQUAL_INT(POLLIN, fds[0].revents);
+
+		conn = accept(fds[0].fd, NULL, NULL);
+		TEST_ASSERT(conn > 0);
+		fds[1].fd = conn;
+		fds[1].events = POLLOUT;
+
+		send_msg(conn, pid, 0);
+
+		if (waitpid(pid, &status, 0) < 0)
+			FAIL("waitpid");
+
+		TEST_ASSERT(WIFEXITED(status));
+		TEST_ASSERT_EQUAL_INT(0, WEXITSTATUS(status));
+
+		close(conn);
+		close(named);
+	}
+	else {
+		if ((fd = socket(AF_UNIX, type, PF_UNSPEC)) < 0)
+			exit(1);
+
+		while (true) {
+			if ((rv = connect_to_named(fd, socket_name)) == 0)
+				break;
+			else {
+				CHILD_ASSERT_ERRNO(rv, ECONNREFUSED);
+				usleep(50);
+			}
+		}
+
+		read_msg(fd, pid, 0);
+
+		close(fd);
+
+		exit(0);
+	}
+
+	/* nonblocking connect, blocking accept */
+
+	pid = checked_fork();
+
+	if (pid) {
```

Comment:
Why are the server and client blocks reversed?

Reply:
There's more stuff to assert in the server block and since asserts are invalid in the child procs these blocks had to be reversed. Otherwise the server would need to call `exit(1)` on failed assertions which isn't informative